<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [1]:
import pandas as pd
import psycopg2
import requests as requests
from bs4 import BeautifulSoup
import lxml

In [3]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [4]:
query = f'''SELECT count(*) FROM vacancies'''

In [5]:
number_of_vacancies = pd.read_sql_query(query, connection)
number_of_vacancies

C:\Users\Антон\AppData\Local\Temp\ipykernel_7292\395214275.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  number_of_vacancies = pd.read_sql_query(query, connection)


count
0  49197

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [6]:
query = f'''SELECT count(*) FROM employers'''

In [7]:
number_of_employers = pd.read_sql_query(query, connection)
number_of_employers

C:\Users\Антон\AppData\Local\Temp\ipykernel_7292\1186626920.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  number_of_employers = pd.read_sql_query(query, connection)


count
0  23501

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [8]:
query = f'''SELECT count(*) FROM areas'''

In [9]:
number_of_areas = pd.read_sql_query(query, connection)
number_of_areas

C:\Users\Антон\AppData\Local\Temp\ipykernel_7292\3513774121.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  number_of_areas = pd.read_sql_query(query, connection)


count
0   1362

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [10]:
query = f'''SELECT count(*) FROM industries'''

In [11]:
number_of_industries = pd.read_sql_query(query, connection)
number_of_industries

C:\Users\Антон\AppData\Local\Temp\ipykernel_7292\3044795537.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  number_of_industries = pd.read_sql_query(query, connection)


count
0    294

***

**Вывод**: у нас имеется 49197 вакансий, 23501 работодатель, 1362 региона и 294 сферы деятельности

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [12]:
query = f'''SELECT
                a.name,
                count(v.*)
            FROM areas a
            JOIN vacancies v ON a.id = v.area_id
            GROUP BY a.name
            ORDER BY 2 DESC
'''

In [13]:
count_vacancies_in_areas = pd.read_sql_query(query, connection)
count_vacancies_in_areas

C:\Users\Антон\AppData\Local\Temp\ipykernel_7292\2537493147.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  count_vacancies_in_areas = pd.read_sql_query(query, connection)


name  count
0             Москва   5333
1    Санкт-Петербург   2851
2              Минск   2112
3        Новосибирск   2006
4             Алматы   1892
..               ...    ...
764       Тарко-Сале      1
765    Новоаннинский      1
766            Бирск      1
767           Сасово      1
768             Анна      1

[769 rows x 2 columns]

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [14]:
query = f'''SELECT
                count(*)
            FROM vacancies
            WHERE salary_from IS NOT NULL OR salary_to IS NOT NULL
'''

In [15]:
query_result = pd.read_sql_query(query, connection)
query_result

C:\Users\Антон\AppData\Local\Temp\ipykernel_7292\1583014093.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  query_result = pd.read_sql_query(query, connection)


count
0  24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [16]:
query = f'''SELECT
                avg(salary_from) avg_from,
                avg(salary_to) avg_to
            FROM vacancies
'''

In [17]:
query_result = pd.read_sql_query(query, connection)
query_result

C:\Users\Антон\AppData\Local\Temp\ipykernel_7292\1583014093.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  query_result = pd.read_sql_query(query, connection)


avg_from         avg_to
0  71064.657901  110536.741923

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [18]:
query = f'''SELECT
                schedule,
                employment,
                count(*)
            FROM vacancies
            GROUP BY schedule, employment
            ORDER BY 3 DESC
'''

In [19]:
query_result = pd.read_sql_query(query, connection)
query_result

C:\Users\Антон\AppData\Local\Temp\ipykernel_7292\1583014093.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  query_result = pd.read_sql_query(query, connection)


schedule           employment  count
0        Полный день     Полная занятость  35367
1   Удаленная работа     Полная занятость   7802
2      Гибкий график     Полная занятость   1593
3   Удаленная работа  Частичная занятость   1312
4     Сменный график     Полная занятость    940
5        Полный день           Стажировка    569
6     Вахтовый метод     Полная занятость    367
7        Полный день  Частичная занятость    347
8      Гибкий график  Частичная занятость    312
9        Полный день     Проектная работа    141
10  Удаленная работа     Проектная работа    133
11     Гибкий график           Стажировка    116
12    Сменный график  Частичная занятость    101
13  Удаленная работа           Стажировка     64
14     Гибкий график     Проектная работа     18
15    Сменный график           Стажировка     12
16    Вахтовый метод     Проектная работа      2
17    Сменный график     Проектная работа      1

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [20]:
query = f'''SELECT
                experience,
                count(*)
            FROM vacancies
            GROUP BY experience
            ORDER BY 2 ASC
'''

In [21]:
query_result = pd.read_sql_query(query, connection)
query_result

C:\Users\Антон\AppData\Local\Temp\ipykernel_7292\1583014093.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  query_result = pd.read_sql_query(query, connection)


experience  count
0         Более 6 лет   1337
1           Нет опыта   7197
2       От 3 до 6 лет  14511
3  От 1 года до 3 лет  26152

***

**Вывод**: большинство вакансий находятся в Москве и Санкт-Петербурге. Примерно у половины вакансий заполнено хотя бы одно из полей нижней и верхней зарплатной вилки. Нижнее значение зарплатной вилки - 71 тысяча рублей, а верхнее - 110 тысяч рублей. Большинство вакансий подразумевают полную занятость и полный день, либо полную занятость и удаленную работу. Около 80% вакансий требуют от 1 года до 6 лет опыта.

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [22]:
query = f'''SELECT
                e.name,
                count(v.*)
            FROM employers e
            JOIN vacancies v ON e.id = v.employer_id
            GROUP BY e.name
            ORDER BY 2 DESC
            LIMIT 5
'''

In [23]:
query_result = pd.read_sql_query(query, connection)
query_result

C:\Users\Антон\AppData\Local\Temp\ipykernel_7292\1583014093.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  query_result = pd.read_sql_query(query, connection)


name  count
0         Яндекс   1933
1     Ростелеком    491
2       Тинькофф    444
3           СБЕР    428
4  Газпром нефть    331

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [24]:
query = f'''SELECT
                a.name,
                count(distinct e.name) employers_count,
                count(distinct v.id) vacancies_count
            FROM areas a
                LEFT JOIN vacancies v ON a.id = v.area_id
                JOIN employers e ON a.id = e.area
            WHERE v.name IS NULL
            GROUP BY a.name
            ORDER BY 2 DESC, 3 DESC
'''

In [25]:
query_result = pd.read_sql_query(query, connection)
query_result

C:\Users\Антон\AppData\Local\Temp\ipykernel_7292\1583014093.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  query_result = pd.read_sql_query(query, connection)


name  employers_count  \
0                                       Россия              410   
1                                    Казахстан              207   
2                           Московская область               75   
3                           Краснодарский край               19   
4                                     Беларусь               18   
..                                         ...              ...   
181                                  Миллерово                1   
182  Михайловка (Уфимский район, Башкортостан)                1   
183                                    Можайск                1   
184                                Морозовск\n                1   
185                                     Несвиж                1   

     vacancies_count  
0                  0  
1                  0  
2                  0  
3                  0  
4                  0  
..               ...  
181                0  
182                0  
183                0  
184                0  
185                0  

[186 rows x 3 columns]

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [26]:
query = f'''SELECT
                e.name,
                count(distinct v.area_id) count_regions
            FROM employers e
                JOIN vacancies v ON v.employer_id = e.id
            GROUP BY e.name
            ORDER BY 2 DESC
'''

In [27]:
query_result = pd.read_sql_query(query, connection)
query_result

C:\Users\Антон\AppData\Local\Temp\ipykernel_7292\1583014093.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  query_result = pd.read_sql_query(query, connection)


name  count_regions
0                      Яндекс            181
1                  Ростелеком            152
2                  Спецремонт            116
3      Поляков Денис Иванович             88
4                   ООО ЕФИН              71
...                       ...            ...
14761                  UniSol              1
14762            UNISTORY LLC              1
14763                   UNIT6              1
14764     United Distribution              1
14765              UNITEL LLC              1

[14766 rows x 2 columns]

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [28]:
query = f'''SELECT
                count(e.name)
            FROM employers e
                LEFT JOIN employers_industries ei ON e.id = ei.employer_id
            WHERE ei.industry_id IS NULL
'''

In [29]:
query_result = pd.read_sql_query(query, connection)
query_result

C:\Users\Антон\AppData\Local\Temp\ipykernel_7292\1583014093.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  query_result = pd.read_sql_query(query, connection)


count
0   8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [30]:
query = f'''SELECT
                e.name
            FROM employers e
                JOIN employers_industries ei ON e.id = ei.employer_id
            GROUP BY e.name
            HAVING count(ei.industry_id) = 4
            ORDER BY e.name
            OFFSET 2 LIMIT 1
'''

In [31]:
query_result = pd.read_sql_query(query, connection)
query_result

C:\Users\Антон\AppData\Local\Temp\ipykernel_7292\1583014093.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  query_result = pd.read_sql_query(query, connection)


name
0  2ГИС

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [32]:
query = f'''SELECT
                count(e.name)
            FROM employers e
                JOIN employers_industries ei ON e.id = ei.employer_id
                JOIN industries i ON ei.industry_id = i.id
            WHERE i.name = 'Разработка программного обеспечения'
'''

In [33]:
query_result = pd.read_sql_query(query, connection)
query_result

C:\Users\Антон\AppData\Local\Temp\ipykernel_7292\1583014093.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  query_result = pd.read_sql_query(query, connection)


count
0   3553

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [34]:
url = 'https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8' # Задаём адрес ресурса
table = pd.read_html(url)[1] # Выбираем вторую таблицу на странице
cities = list(table['Город'])
cities.remove('Воронеж[a]')
cities.append('Воронеж')
cities = tuple(cities)

In [35]:
query = f'''(SELECT
              a.name,
              count(v.*)
            FROM vacancies v
                JOIN areas a ON a.id = v.area_id
                JOIN employers e ON e.id = v.employer_id
            WHERE a.name IN {cities} AND e.name = 'Яндекс'
            GROUP BY 1
            )

            UNION ALL

            (SELECT
                'Total',
                count(v.*)
            FROM vacancies v
                JOIN areas a ON a.id = v.area_id
                JOIN employers e ON e.id = v.employer_id
            WHERE a.name IN {cities} AND e.name = 'Яндекс'
            )

            ORDER BY 2 ASC
'''

In [36]:
query_result = pd.read_sql_query(query, connection)
query_result

C:\Users\Антон\AppData\Local\Temp\ipykernel_7292\1583014093.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  query_result = pd.read_sql_query(query, connection)


name  count
0              Омск     21
1         Челябинск     22
2        Красноярск     23
3         Волгоград     24
4             Пермь     25
5            Казань     25
6    Ростов-на-Дону     25
7               Уфа     26
8            Самара     26
9         Краснодар     30
10          Воронеж     32
11      Новосибирск     35
12  Нижний Новгород     36
13     Екатеринбург     39
14  Санкт-Петербург     42
15           Москва     54
16            Total    485

***

**Вывод**: самое большое количество вакансий предлагают крупные корпорации - Яндекс, Ростелеком, Тинькофф, СБЕР и Газпром нефть. Яндекс и Ростелеком - лидеры по количеству регионов, в которых размещаются вакансии. Почти у трети работодателей не указана сфера деятельности. Примерно у 15% работодателей сфера деятельности - это разработка программного обеспечения. Большинство вакансий Яндекса открыты в Москве и Санкт-Петербурге.


# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [37]:
query = f'''SELECT
                count(*)
            FROM vacancies v
            WHERE v.name ILIKE '%data%' OR v.name ILIKE '%данн%'
'''

In [38]:
query_result = pd.read_sql_query(query, connection)
query_result

C:\Users\Антон\AppData\Local\Temp\ipykernel_7292\1583014093.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  query_result = pd.read_sql_query(query, connection)


count
0   1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [39]:
query = f'''SELECT
                count(*)
            FROM vacancies v
            WHERE
                (v.name ILIKE '%data%scientist%' OR
                v.name ILIKE '%data%science%' OR
                v.name ILIKE '%исследователь%данных%' OR
                (v.name LIKE '%ML%' AND v.name NOT ILIKE '%HTML%') OR
                v.name ILIKE '%machine%learning%' OR
                v.name ILIKE '%машинн%обучен%')
                AND
                (v.name ILIKE '%junior%' OR
                v.experience ILIKE '%Нет опыта%' OR
                v.employment ILIKE '%Стажировка%')
'''

In [40]:
query_result = pd.read_sql_query(query, connection)
query_result

C:\Users\Антон\AppData\Local\Temp\ipykernel_7292\1583014093.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  query_result = pd.read_sql_query(query, connection)


count
0     51

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [41]:
query = f'''SELECT
                count(*)
            FROM vacancies v
            WHERE
                (v.name ILIKE '%data%scientist%' OR
                v.name ILIKE '%data%science%' OR
                v.name ILIKE '%исследователь%данных%' OR
                (v.name LIKE '%ML%' AND v.name NOT ILIKE '%HTML%') OR
                v.name ILIKE '%machine%learning%' OR
                v.name ILIKE '%машинн%обучен%')
                AND
                (v.key_skills ILIKE '%SQL%' OR
                v.key_skills ILIKE '%postgres%')
'''

In [42]:
query_result = pd.read_sql_query(query, connection)
query_result

C:\Users\Антон\AppData\Local\Temp\ipykernel_7292\1583014093.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  query_result = pd.read_sql_query(query, connection)


count
0    201

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [43]:
query = f'''SELECT
                count(*)
            FROM vacancies v
            WHERE
                (v.name ILIKE '%data%scientist%' OR
                v.name ILIKE '%data%science%' OR
                v.name ILIKE '%исследователь%данных%' OR
                (v.name LIKE '%ML%' AND v.name NOT ILIKE '%HTML%') OR
                v.name ILIKE '%machine%learning%' OR
                v.name ILIKE '%машинн%обучен%')
                AND
                v.key_skills ILIKE '%Python%'
'''

In [44]:
query_result = pd.read_sql_query(query, connection)
query_result

C:\Users\Антон\AppData\Local\Temp\ipykernel_7292\1583014093.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  query_result = pd.read_sql_query(query, connection)


count
0    351

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [45]:
query = f'''SELECT
                avg(character_length(v.key_skills) - character_length(replace(v.key_skills, CHR(9), ''))+1)
            FROM vacancies v
            WHERE
                (v.name ILIKE '%data%scientist%' OR
                v.name ILIKE '%data%science%' OR
                v.name ILIKE '%исследователь%данных%' OR
                (v.name LIKE '%ML%' AND v.name NOT ILIKE '%HTML%') OR
                v.name ILIKE '%machine%learning%' OR
                v.name ILIKE '%машинн%обучен%')
                AND v.key_skills IS NOT NULL
'''

In [46]:
query_result = pd.read_sql_query(query, connection)
query_result

C:\Users\Антон\AppData\Local\Temp\ipykernel_7292\1583014093.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  query_result = pd.read_sql_query(query, connection)


avg
0  6.408776

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [47]:
query = f'''SELECT
                v.experience,
                avg((coalesce(v.salary_from, v.salary_to) + coalesce(v.salary_to, v.salary_from))/2)
            FROM vacancies v
            WHERE
                (v.name ILIKE '%data%scientist%' OR
                v.name ILIKE '%data%science%' OR
                v.name ILIKE '%исследователь%данных%' OR
                (v.name LIKE '%ML%' AND v.name NOT ILIKE '%HTML%') OR
                v.name ILIKE '%machine%learning%' OR
                v.name ILIKE '%машинн%обучен%')
                AND
                (v.salary_from IS NOT NULL OR v.salary_to IS NOT NULL)
            GROUP BY 1
'''

In [48]:
query_result = pd.read_sql_query(query, connection)
query_result

C:\Users\Антон\AppData\Local\Temp\ipykernel_7292\1583014093.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  query_result = pd.read_sql_query(query, connection)


experience            avg
0           Нет опыта   74642.857143
1  От 1 года до 3 лет  139674.750000
2       От 3 до 6 лет  243114.666667

***

**Вывод**: всего 1771 вакансия из примерно 49 тысяч имеет отношение к данным. Для начинающих дата-сайентистов имеется всего 51 вакансия. 201 вакансия требует знания SQL или PostgreSQL. В среднем вакансии требуют 6-7 ключевых навыков. В вакансиях без опыта средняя предлагаемая заработная плата около 75 тысяч; при опыте работы от 1 года до 3 лет - около 140 тысяч; при опыте от 3 до 6 лет - около 243 тысяч.

# Общий вывод по проекту

In [49]:
# У какой компании-работодателя больше всего сфер деятельности
query = f'''SELECT
                e.name,
                count(distinct i.name)
            FROM employers e
                JOIN employers_industries ei ON e.id = ei.employer_id
                JOIN industries i ON ei.industry_id = i.id
            GROUP BY 1
            ORDER BY 2 DESC
'''

In [50]:
query_result = pd.read_sql_query(query, connection)
query_result

C:\Users\Антон\AppData\Local\Temp\ipykernel_7292\1583014093.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  query_result = pd.read_sql_query(query, connection)


name  count
0                          Модуль     16
1                         Энергия     10
2                           Прайд     10
3                             СБК      9
4                           Альфа      9
...                           ...    ...
14951             ПассатИнновации      1
14952     Хьюман Аппс/ Human Apps      1
14953          Хьюман Фокус Медиа      1
14954                     Divo.ai      1
14955  ПАУЭР ИНЖИНИРИНГ Казахстан      1

[14956 rows x 2 columns]

In [51]:
# Где больше всего вакансий с удаленной работой
query = f'''(SELECT
                e.name,
                count(v.*)
            FROM employers e
                JOIN vacancies v ON e.id = v.employer_id
            WHERE v.schedule = 'Удаленная работа'
            GROUP BY 1
            ORDER BY 2 DESC)

            UNION ALL

            (SELECT
                'Total',
                count(v.*)
            FROM employers e
                JOIN vacancies v ON e.id = v.employer_id
            WHERE v.schedule = 'Удаленная работа'
            GROUP BY 1
            ORDER BY 2 DESC)
'''

In [52]:
query_result = pd.read_sql_query(query, connection)
query_result

C:\Users\Антон\AppData\Local\Temp\ipykernel_7292\1583014093.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  query_result = pd.read_sql_query(query, connection)


name  count
0                       Total   9311
1                      Яндекс   1690
2                     DataArt    165
3                    Тинькофф     95
4          Nitka Technologies     77
...                       ...    ...
2326             Мегаптека.ру      1
2327                   Hexlet      1
2328                Globus.ai      1
2329  Медиагруппа FM-Продакшн      1
2330               Информбокс      1

[2331 rows x 2 columns]

**Вывод**: всего в данных таблицах представлено около 49 тысяч вакансий и 23.5 тысыч работодателей. Большое количество вакансий распределено в городах с большим населением (например Москва и Санкт-Петербург). Можно заметить что в половине вакансий не заполнена нижняя либо верхняя границы зарплатной вилки. В основном работодатели ждут в своих компаниях работников на полный рабочий день с полной занятостью, так же некоторую весомую часть имеют вакансии, позволяющие работать удаленно. На половину вакансий готовы принимать с опытом от года до трёх лет; для соискателей найти работу без опыта тяжелее. Из дополнительных исследований можно выяснить, что многие компании имеют 1-2 сферы деятельности, но так же, например, имеется компания "Модуль", у которых их целых 16. У трети работодателей не указана сфера деятельности. Разработка ПО как сфера деятельности указана примерно у 15% работодателей. Отношение к данным имеет около 3.5% вакансий. Для DS'a уровня junior есть 51 вакансия. Во многих вакансиях для DS знание SQL является одним из ключевых навыков, как и знание Python. Среднее значение ключевых навыков для DS'а колеблется от 6 до 7. В зависимости от опыта работы (категорий "Нет опыта", "1-3 года", "3-6 лет") работодатели готовы предложить намного большую оплату труда. Большим приоритетом для многих людей является удаленная работа: компания "Яндекс" лидирует по количеству такого типа вакансий. Это хороший проект, который может помочь начинающим DS'ам, но не людям с большим опытом работы, так как я думаю, что они имеют большее понимание области своей работы, а так же области поиска работы. Для продолжения проекта можно брать данные не только с сайта HeadHunter, но и с других сайтов, предназначенных для поиска работы.